server/blockchain/calypso/calypso.go

```diff
@@ -491,3 +590,151 @@ func (h infoLog) Write(p []byte) (int, error) {
+		Msgf("setting %x=%s", key, roster)
+
+	return nil
+}
+
+func getSecret(snap store.Snapshot, key []byte) ([]byte, error) {
+	k := prefixed.NewPrefixedKey([]byte(PrefixSecretKeys), key)
+	secret, err := snap.Get(k)
+	if err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+func setSecret(snap store.Snapshot, key []byte, secret []byte) error {
+	k := prefixed.NewPrefixedKey([]byte(PrefixSecretKeys), key)
+	err := snap.Set(k, secret)
+	if err != nil {
+		return err
+	}
+
+	dela.Logger.Info().
+		Str("contract", ContractName).
+		Msgf("setting secret %x=%s", key, secret)
+
+	return nil
+}
+
+func deleteSecret(snap store.Snapshot, key []byte) error {
+	k := prefixed.NewPrefixedKey([]byte(PrefixSecretKeys), key)
+	err := snap.Delete(k)
+	if err != nil {
+		return xerrors.Errorf("failed to delete from snapshot: %v", err)
+	}
+
+	dela.Logger.Info().
+		Str("contract", ContractName).
+		Msgf("deleting secret %x", key)
+
+	return nil
+}
+
+func computeAccessToken(smcKey []byte, secret []byte, clientPubKey []byte) []byte {
+	h := crypto.NewHashFactory(crypto.Sha3_224).New()
+	h.Write(smcKey)
+	h.Write(secret)
+	h.Write(clientPubKey)
+	return h.Sum(nil)
+}
+
+func hasSecretAccess(snap store.Snapshot, accessToken []byte) bool {
+	_, err := getSecret(snap, accessToken)
+	return err != nil
+}
+
+func getSecretAccess(snap store.Snapshot, accessToken []byte) ([]byte, error) {
+	k := prefixed.NewPrefixedKey([]byte(PrefixAccessKeys), accessToken)
+	pubKey, err := snap.Get(k)
+	if err != nil {
+		return nil, xerrors.Errorf(
+			"failed to get access token '%v': %v", accessToken, err)
+	}
+
+	return pubKey, nil
+}
+
+func setSecretAccess(snap store.Snapshot, accessToken []byte, pubKey []byte) error {
+	k := prefixed.NewPrefixedKey([]byte(PrefixAccessKeys), accessToken)
+	err := snap.Set(k, pubKey)
+	if err != nil {
+		return xerrors.Errorf(
+			"failed to give secret '%v' access to '%v': %v", accessToken, pubKey, err)
+	}
+
+	dela.Logger.Info().
+		Str("contract", ContractName).
+		Msgf("setting secret access %x=%s", accessToken, pubKey)
+
+	return nil
+}
+
+func insertAuditLog(snap store.Snapshot, name []byte, accessToken []byte) error {
+	k := prefixed.NewPrefixedKey([]byte(PrefixListKeys), name)
+	log, err := snap.Get(k)
+	if err != nil {
+		log = make([]byte, 0, 28)
+	}
+
+	// log contains: [<accessToken1>, <accessToken2>, ...] , [][]byte
+	log = append(log, accessToken...)
+
+	err = snap.Set(k, log)
+	if err != nil {
+		return xerrors.Errorf(
+			"failed to insert audit log for secret '%v': %v", name, err)
+	}
+
+	dela.Logger.Info().
+		Str("contract", ContractName).
+		Msgf("appending audit log %x=[%s]", name, accessToken)
+
+	return nil
+}
+
+func getAuditLogs(snap store.Snapshot, name []byte) ([][]byte, error) {
+	k := prefixed.NewPrefixedKey([]byte(PrefixListKeys), name)
+	log, err := snap.Get(k)
+	if err != nil {
+		return [][]byte{}, err
+	}
+
+	return decodeAuditLog(log)
+}
+
+func decodeAuditLog(log []byte) ([][]byte, error) {
```

Comment:
if we move back to SHA-256, the 28-bytes-long logic doesn't make sense anymore.

Also, we use the number `28` a ridiculous amount of times. Introducing a constant wouldn't hurt. (My bad + GitHub Copilot's, I know)